Feature/add consultancy account

GustaafL opened the pull request:
## Description

Add a Consultant account that has read access to accounts of clients. 

Use case: Our customer is the supplier to a bunch of other (client) accounts on the server. They should have read-only access to some accounts. At least this is the first step. So they can oversee what happens, for instance for servicing questions and problems.

## Look & Feel

The consultant will be able to see the accounts in their overview page and the assets they have access to. The consultant client will be able to see which account has read access to their account.

## How to test

Tests are added in flexmeasures/api/v3_0/tests/test_assets_api.py 
To test in the UI create a consultant account and a consultant client account in the CLI. Then go to the account pages and see the accounts represented.

## Further Improvements

Potential improvements to be done in the same PR or follow up Issues/Discussions/PRs.

## Related Items

The technical discussion can be found here: https://github.com/FlexMeasures/flexmeasures/discussions/868
The issue that this closes can be found here: #203

---

- [x] I agree to contribute to the project under Apache 2 License. 
- [x] To the best of my knowledge, the proposed patch is not based on code under GPL or other license that is incompatible with FlexMeasures

nhoening commented on documentation/changelog.rst:
```diff
@@ -13,6 +13,7 @@ New features
 -------------
 
 - Scheduling data better distinguishes (e.g. in chart tooltips) when a schedule was the result of a fallback mechanism, by splitting off the fallback mechanism from the main scheduler (as a separate job) [see `PR #846 <https://github.com/FlexMeasures/flexmeasures/pull/846>`_]
+- A Consultancy account can be added that has read access to accounts its id in the `consultancy_account_id` column. [see `PR #846 <https://github.com/FlexMeasures/flexmeasures/pull/877>`_]
```
"A Consultancy account can be added to any account, with the effect that the users in that account get read access."

GustaafL replied:
Thanks, changed to suggested

nhoening commented on flexmeasures/api/v3_0/tests/test_sensors_api.py:
```diff
@@ -28,7 +28,6 @@ def test_fetch_one_sensor(
     assert response.status_code == 200
     assert response.json["name"] == "some gas sensor"
     assert response.json["unit"] == "m³/h"
-    assert response.json["generic_asset_id"] == 4
```
Any idea why this is not returned anymore?

GustaafL replied:
Great question! The asset id is not explicitly assigned anywhere but is created when the asset is created. This means it's subject to change when changes are made to conftest. Because of this the test could break with future changes as well and changing the number in this test would not actually make this test valuable. 

I've changed the test to first get the generic_asset linked to this sensor and check that it's linked to the "incineration line" asset.

nhoening commented on flexmeasures/conftest.py:
```diff
@@ -156,6 +162,23 @@ def create_test_accounts(db) -> dict[str, Account]:
     )
 
 
+def add_consult_account_access_to_prosumer_account(db, test_accounts):
```
It seems this logic could just be inside `create_test_accounts()`, as it is used in both functions that call it - and it creates an account.

GustaafL replied:
updated, removed the extra function and added the function to `create_test_accounts()`

nhoening commented on flexmeasures/api/v3_0/tests/test_assets_api.py:
```diff
@@ -359,3 +359,30 @@ def test_delete_an_asset(client, setup_api_test_data, requesting_user):
+    requesting_user,
+):
+    """
+    Get assets in the account the Prosumer account is in the
```
This sentence is too difficult to read :) Make two out if it of needed.

nhoening commented on flexmeasures/conftest.py:
```diff
@@ -109,12 +109,18 @@ def create_test_db(app):
 def setup_accounts_fresh_db(fresh_db) -> dict[str, Account]:
-    return create_test_accounts(fresh_db)
+    test_accounts = create_test_accounts(fresh_db)
+    test_accounts |= add_consult_account_access_to_prosumer_account(
```
This operator is not supported in Python 3.8, see https://github.com/FlexMeasures/flexmeasures/actions/runs/6549108023/job/17785337960?pr=877

Anyway, I think we don't need to do it like this, see my other comment. Then this problem would also go away.

GustaafL replied:
fixed `create_test_accounts()` to include the consultanct account and therefore was able to change this function back to the original version. This also fixes Python 3.8 support

nhoening commented on flexmeasures/conftest.py:
```diff
@@ -156,6 +162,23 @@ def create_test_accounts(db) -> dict[str, Account]:
+    account = Account.query.filter_by(name="Test Prosumer Account").one_or_none()
+    consultant_account_role = AccountRole(
+        name="Consultant",
+        description="Account that prosumer can access",
```
This is the wrong way around. The consultant can access accounts of its clients (which in your case here is the Prosumer). Please reverse this and your test.

GustaafL replied:
Changed the `conftest` by adding a ConsultantClient account with an asset. The Consultant account has access to the ConsultantClient account.

nhoening commented on flexmeasures/api/v3_0/tests/test_assets_api.py:
```diff
@@ -359,3 +359,30 @@ def test_delete_an_asset(client, setup_api_test_data, requesting_user):
+    ["test_prosumer_user@seita.nl"],
+    indirect=True,
+)
+def test_get_assets_new(
```
What does the "new" signify here? Would not `test_consultant_can_read` be better?

GustaafL replied:
Thanks for catching this. changed naming to `test_consultant_can_read`.

nhoening commented on flexmeasures/api/v3_0/tests/test_assets_api.py:
```diff
@@ -359,3 +359,30 @@ def test_delete_an_asset(client, setup_api_test_data, requesting_user):
+    )
+    print("Server responded with:\n%s" % get_assets_response.json)
+    assert get_assets_response.status_code == 200
+    assert len(get_assets_response.json) == 1
```
There should also be a test showing that the consultant cannot do things other than read, e.g. using the patch endpoint.

GustaafL replied:
Added test to show the assets can not be patched by a consultant

nhoening reviewed (changes requested):
Good start!

I found a mix-up in how it's supposed to work, though.

Can you edit the PR description (only relevant sections), and then also reference discussion #868 

Once we have this working well, I think I'd like to implement my idea of the required user role, as well (see discussion: "Addendum: also require a user role"). Better to do that right away, because otherwise the access might apply to too many people.

GustaafL commented:
I've added the customer-manager user role as a requirement for accessing consultant client accounts by consultants.

nhoening commented on flexmeasures/api/v3_0/tests/test_assets_api.py:
```diff
@@ -366,3 +365,113 @@ def test_get_assets_new(
+    asset_edit_response = client.patch(
+        url_for("AssetAPI:patch", id=consultant_client_asset.id),
+        json={
+            "latitude": consultant_client_asset.latitude,
```
More simply, set to a bogus value like 0. Reads easier (in the sense that we actually attempt to change the value)

GustaafL replied:
Changed the value to 0 so it's more clear the value is not important here.

nhoening commented on flexmeasures/api/v3_0/tests/test_assets_api.py:
```diff
@@ -366,3 +365,113 @@ def test_get_assets_new(
+        ("parent", "child_1", True),
+    ],
+)
+@pytest.mark.parametrize("requesting_user", ["test_admin_user@seita.nl"], indirect=True)
```
Something went wrong when merging, because this test comes from [another PR](https://github.com/FlexMeasures/flexmeasures/pull/874/files#diff-045dbe938b968f5c33e77033a57045cdcf6e24f77b7b466405aa391ecdff624eR344).

So this should go.

I'm not sure if there are more like that, can you check?

GustaafL replied:
Thanks for pointing this out. Apparently it is correct but good to check the merge because it was indeed a confusing one. This line is also in line 353 of the other PR.

nhoening commented on flexmeasures/conftest.py:
```diff
@@ -246,12 +244,27 @@ def create_roles_users(db, test_accounts) -> dict[str, User]:
+            password="testtest",
+        )
+    )
+    new_users.append(
```
This guy is not being used I believe.

GustaafL replied:
removed this unused user

nhoening commented on flexmeasures/conftest.py:
```diff
@@ -154,28 +148,32 @@ def create_test_accounts(db) -> dict[str, Account]:
+        description="A Client of a Consultant",
+    )
+    account = Account.query.filter_by(name="Test Consultant Account").one_or_none()
+    consultant_client_account = Account(
```
I believe we don't need to add this account. We could give the consultant the PRosumer as a client, as we're not doing anything invasive anyway. Would also spare us another asset.

```
prosumer_account.consultancy_account_id = account.id
```

(maybe `account` is a variable name which can lead to confusions btw)

GustaafL replied:
I've removed the account variable in favor of just the id with a more descriptive name: `consultant_account_id = Account.query.filter_by(name="Test Consultant Account").one_or_none().id`

I think the extra consultant client account makes it easier to see the relationship for future reference. I think the Prosumer name can also be confusing for being a consultant client account.

I can change it to test on the Prosumer if needed but I'm not sure it's worth changing this in terms of resources or making the test setup more clear.

nhoening reviewed (changes requested):
Good iteration! Only smaller comments at this point.

Maybe we should add a small sentence/paragraph to https://flexmeasures.readthedocs.io/en/latest/concepts/security_auth.html#authorization, to explain this concept.

GustaafL commented:
I've added the customer-manager user role as a requirement for accessing consultant client accounts by consultants.

I've also updated the description of the PR.

nhoening reviewed (changes requested):
Thanks!
 I found a few ways to improve.
 
 Also, I wanted to also add a note to the documentation, see my last review.

GustaafL commented on flexmeasures/data/models/user.py:
```diff
@@ -65,6 +65,17 @@ class Account(db.Model, AuthModelMixin):
+    consultancy_account_id = Column(
+        Integer, db.ForeignKey("account.id"), default=None, nullable=True
+    )
+
```
Moved the blank line to make a clear distinction where the consultant account modeling is done.

GustaafL commented on flexmeasures/data/services/accounts.py:
```diff
@@ -10,16 +10,27 @@ def get_accounts(
     The role_name parameter allows to filter by role.
     """
-    account_query = Account.query
+    # account = Account
```
removed the commented out code.

GustaafL commented on flexmeasures/data/models/user.py:
```diff
@@ -75,9 +86,15 @@ def __acl__(self):
         Within same account, everyone can read and update.
```
added documentation to state that consultant accounts can read in accounts they are consultants for.

GustaafL commented on flexmeasures/ui/templates/defaults.jinja:
```diff
@@ -34,14 +34,20 @@
+{% set view = "non_admin" if current_user.is_authenticated and not (current_user.has_role('admin') or FLEXMEASURES_MODE == "demo" or current_user.has_role('customer-manager')) %}
+{% do navigation_bar.append(("accounts/{}".format(current_user.account.id), 'accounts', 'My Account', '', 'cubes')) if view == "non_admin" %}
+
+{% set view = "consultant" if current_user.is_authenticated and (current_user.has_role('customer-manager') or FLEXMEASURES_MODE == "demo") %}
```
Changed the views to the correct logic and manually tested the different types of views.

GustaafL commented on flexmeasures/ui/templates/defaults.jinja:
```diff
@@ -34,14 +34,20 @@
+
+{% set view = "consultant" if current_user.is_authenticated and (current_user.has_role('customer-manager') or FLEXMEASURES_MODE == "demo") %}
+{% do navigation_bar.append(('accounts', 'accounts', 'Accounts', '', 'cubes')) if view == "consultant" %}
+{% do navigation_bar.append(('assets', 'assets', 'Assets', '', 'list-ul')) if view == "consultant" %}
```
removed the extra views that consultants should not have.

GustaafL commented on flexmeasures/ui/crud/accounts.py:
```diff
@@ -46,6 +47,13 @@ def get(self, account_id: str):
+            consultancy_account_name = (
+                Account.query.filter_by(id=account["consultancy_account_id"])
+                .one_or_none()
+                .name
```
changed to make this not fail if there is a consultany account id but not an consultant account

nhoening reviewed (changes requested):
Very close to done.
I'd like to add the consultant's account ID to the client account when I'm creating it on the CLI (we discussed that once in person).
That would also make it easy enough for me to test this PR locally.

GustaafL commented on documentation/changelog.rst:
```diff
@@ -13,6 +13,7 @@ New features
 -------------
 
 - Scheduling data better distinguishes (e.g. in chart tooltips) when a schedule was the result of a fallback mechanism, by splitting off the fallback mechanism from the main scheduler (as a separate job) [see `PR #846 <https://github.com/FlexMeasures/flexmeasures/pull/846>`_]
+- A Consultancy account can be added to any account, with the effect that the users in that account get read access. [see `PR #846 <https://github.com/FlexMeasures/flexmeasures/pull/877>`_]
```
added that customer-manager users get read access.

GustaafL commented on flexmeasures/ui/templates/crud/account.html:
```diff
@@ -29,6 +29,16 @@ <h3>Account</h3>
+                        {% if account.consultancy_account_name %}
+                        <tr>
+                            <td>
+                               Your Consultant
```
changed to Consultant

GustaafL commented:
I've added the option to add a consultancy-account-id in the CLI and a test that this works as expected.

nhoening commented on flexmeasures/ui/templates/defaults.jinja:
```diff
@@ -34,14 +34,17 @@
+{% do navigation_bar.append(('users', 'users', 'Users', '', 'users')) if view == "admin" %}
+{% do navigation_bar.append(('tasks', 'tasks', 'Tasks', '', 'tasks')) if view == "admin" %}
+
+{% set view = "non_admin" if current_user.is_authenticated and not (current_user.has_role('admin') or FLEXMEASURES_MODE == "demo" or current_user.has_role('customer-manager')) %}
```
Difficult to read. Let's stay as atomic and I would use separate flags, not just `view`.
Maybe `is_admin` and `is_consultant` (one can be both).

I guess `non_admin` is not a good term here - and I would not use it.

We already have  `admin` and `consultant` -  "My account" should be shown if the user is authenticated, is neither admin nor consultant. And apparently when not in demo mode either?

GustaafL replied:
Refactored the code. The original was quite confusing as well, I think the current version is the most clear and can be understood and changed.

nhoening reviewed (changes requested):
I tested it (via the UI) and it works.
I feel that one part is too confusing for later, and needs more clarity. Then we can merge!

Flix6x commented:
I used these psql statements to get a short demo going. Perhaps the first statement is suited to already be included in the database migration.
```
insert into role (name, description) values ('customer-manager', 'Consultant that manages another customer');
insert into roles_users (user_id, role_id) values (<user ID of consultant>, <role ID that was just created>);
update account set consultancy_account_id=<account ID of consultant organisation> where id=<account ID of consultee organisation>;
```

nhoening commented:
So we should add two new roles there in our current PR? Thanks.

GustaafL commented:
> So we should add two new roles there in our current PR? Thanks.

Done

nhoening reviewed (approved):
Nice work!
